File: frontend/lib/constants/constructionSubcategory.ts
// UAT 2026-09-07: "Produksi" tunggal tidak cukup untuk menentukan HPP —
// Produksi Subsidi dan Komersial WAJIB terpisah karena nilai biaya
// produksinya berbeda dan alokasinya HANYA jatuh ke unit dengan tax_category
// yang sama (lihat allocation.ComputeHardPool di backend). Sarana & Prasarana
// dan Perizinan tetap dialokasikan sesuai rule existing (lintas semua unit).
//
// 4 nilai ini adalah string kanonik domain.ConstructionSubcategory — HARUS
// tetap sinkron dengan backend/internal/domain/construction_subcategory.go.
// Dipakai baik di RAB (BudgetItem.subcategory) maupun Cost Entry
// (CostEntry.hard_subcategory) — satu sumber kebenaran di frontend untuk
// keduanya supaya label dan nilai tidak pernah bergeser antar form.
export const CONSTRUCTION_SUBCATEGORIES: { value: string; label: string }[] = [
  { value: "produksi_subsidi",   label: "Produksi — Subsidi" },
  { value: "produksi_komersial", label: "Produksi — Komersial" },
  { value: "sarana_prasarana",   label: "Sarana & Prasarana" },
  { value: "perizinan",          label: "Perizinan" },
];

export function constructionSubcategoryLabel(value: string): string {
  return CONSTRUCTION_SUBCATEGORIES.find(s => s.value === value)?.label ?? value;
}

File: backend/internal/domain/construction_subcategory.go
package domain

// ConstructionSubcategory memecah CostCategoryHard (Konstruksi) menjadi 4
// klasifikasi yang WAJIB terpisah untuk kebutuhan HPP (UAT 2026-09-07,
// business rule final klien):
//
//   - Produksi Subsidi dan Produksi Komersial nilainya berbeda dan TIDAK BOLEH
//     dicampur dalam satu pool alokasi — masing-masing hanya boleh membentuk
//     HPP unit/blok dengan TaxCategory yang sama (lihat RestrictsToTaxCategory).
//   - Sarana & Prasarana dan Perizinan tetap dialokasikan ke SEMUA unit
//     HPP-eligible sesuai basis alokasi existing (tidak dibatasi TaxCategory).
//
// SSOT — jangan sebarkan literal string subkategori di package lain.
type ConstructionSubcategory string

const (
	ConstructionProduksiSubsidi   ConstructionSubcategory = "produksi_subsidi"
	ConstructionProduksiKomersial ConstructionSubcategory = "produksi_komersial"
	ConstructionSaranaPrasarana   ConstructionSubcategory = "sarana_prasarana"
	ConstructionPerizinan         ConstructionSubcategory = "perizinan"
)

// AllConstructionSubcategories adalah urutan kanonik 4 subkategori Konstruksi.
var AllConstructionSubcategories = []ConstructionSubcategory{
	ConstructionProduksiSubsidi,
	ConstructionProduksiKomersial,
	ConstructionSaranaPrasarana,
	ConstructionPerizinan,
}

func (c ConstructionSubcategory) Valid() bool {
	switch c {
	case ConstructionProduksiSubsidi, ConstructionProduksiKomersial, ConstructionSaranaPrasarana, ConstructionPerizinan:
		return true
	default:
		return false
	}
}

// Label mengembalikan nama tampilan subkategori — SSOT untuk teks yang dilihat
// pengguna (RAB vs Realisasi Konstruksi), supaya tidak ada package lain yang
// menuliskan ulang label ini sebagai string literal.
func (c ConstructionSubcategory) Label() string {
	switch c {
	case ConstructionProduksiSubsidi:
		return "Produksi Subsidi"
	case ConstructionProduksiKomersial:
		return "Produksi Komersial"
	case ConstructionSaranaPrasarana:
		return "Sarana & Prasarana"
	case ConstructionPerizinan:
		return "Perizinan"
	default:
		return string(c)
	}
}

// RestrictsToTaxCategory mengembalikan TaxCategory yang berhak menerima HPP
// dari subkategori ini, dan true bila subkategori ini memang dibatasi.
// Sarana & Prasarana dan Perizinan mengembalikan (_, false) — dialokasikan ke
// semua unit HPP-eligible sesuai rule existing, tidak dibatasi Subsidi/Komersial.
func (c ConstructionSubcategory) RestrictsToTaxCategory() (TaxCategory, bool) {
	switch c {
	case ConstructionProduksiSubsidi:
		return TaxCategorySubsidi, true
	case ConstructionProduksiKomersial:
		return TaxCategoryKomersial, true
	default:
		return "", false
	}
}
